### PR TITLE
Improve the stability of measures on the GPU.

### LINKED ISSUE #85: Stabilize measures on the GPU when the CPU is limiting
When the CPU is limiting and the GPU underused, measures are innacurate: #84 shows a difference that can reach a factor of 2 between the measured time when the CPU is limiting compared to when the GPU is limiting.

Ideally, the fix should not impact performance: more evaluations are needed only in the case where the CPU was limiting. Bellow are multiple ideas to have stable measures:
- launch dummy kernel when the GPU is not working to keep it busy
- find another approach to pin the GPU clock
  - not available on ficus
- use events to detect when the measured and actual time do not match
  - some incorrect time may not be detected if the two are too close
- lauch each kernels multiple times.
  - exit early if the perf is too bad

### PR DESCRIPTION
**Pull request by ulysseB**
This PR fixes #85.

It makes the cuda context run each kernel 20 times if the first evaluation is within 3 times the execution time of the best candidate. The `cuda_variance` benchmark shows this is enough to have a delta inferior at 1% on 50 calls to the evaluation method.

Better accuracy at a much smaller cost is achievable by exiting after a few evaluation if the execution time is not good enough. This will be done in a future PR.

### CODE REVIEW

File: src/device/cuda/context.rs [hunk 1]
```diff
@@ -60,6 +70,40 @@ impl<'a> Context<'a> {
...
+        let t0 = self.ticks_to_ns(unwrap!(thunk.execute()));
+        if mode.skip_bad_candidates() && t0 * SKIP_THRESHOLD >= bound {
+            info!("candidate skipped after its first evaluation");
+            return Ok(t0);
+        }
+        // TODO(cc_perf): becomes the limiting factor after a few hours. We should stop
+        // earlier and make tests to know when (for example, measure the MAX delta between
+        // min and median with N outliers).
+        let runtimes = (0..NUM_EVALS).map(|_| thunk.execute());
+        let runtimes_by_value = process_results(runtimes, |iter| iter.sorted())?;
```
Comment (Elarnon) [inline src/device/cuda/context.rs]: Is `process_results` defined somewhere?

File: src/device/cuda/context.rs [hunk 2]
```diff
@@ -60,6 +70,40 @@ impl<'a> Context<'a> {
...
+        }
+        let t0 = self.ticks_to_ns(unwrap!(thunk.execute()));
+        if mode.skip_bad_candidates() && t0 * SKIP_THRESHOLD >= bound {
+            info!("candidate skipped after its first evaluation");
+            return Ok(t0);
+        }
+        // TODO(cc_perf): becomes the limiting factor after a few hours. We should stop
+        // earlier and make tests to know when (for example, measure the MAX delta between
+        // min and median with N outliers).
+        let runtimes = (0..NUM_EVALS).map(|_| thunk.execute());
```
Comment (Elarnon) [inline src/device/cuda/context.rs]: Does this mean the kernel will get added `NUM_EVALS` times to the execution queue after the previous one has finished (i.e. they will not be consecutive)? Not that I'm sure whether this is a good or a bad thing...

Comment (ulysseB) [reply src/device/cuda/context.rs]: No. `thunk.executes()` directly executes the kernel and returns its execution time. This code is called by the code that consumes the execution queue.

File: src/device/cuda/context.rs (hunk 1 above)
Comment (ulysseB) [reply src/device/cuda/context.rs]: It is imported by `use itertools::*;`. I will make the import explicit.

File: src/device/cuda/context.rs (hunk 2 above)
Comment (Elarnon) [reply src/device/cuda/context.rs]: Ah I see

**Pull request merged**